test/common/ssl/ssl_socket_test.cc

```diff
@@ -54,19 +54,118 @@ namespace Ssl {
 
 namespace {
 
-// TODO replace the long parameter list with an options object
-void testUtil(const std::string& client_ctx_yaml, const std::string& server_ctx_yaml,
-              const std::string& expected_digest, const std::string& expected_uri,
-              const std::string& expected_local_uri, const std::string& expected_serial_number,
-              const std::string& expected_subject, const std::string& expected_local_subject,
-              const std::string& expected_peer_cert, const std::string& expected_stats,
-              bool expect_success, const Network::Address::IpVersion version) {
+/**
+ * A base class to hold the options for testUtil() and testUtilV2().
+ */
+class TestUtilOptionsBase {
+public:
+  const std::string& expectedClientCertURI() const { return expected_client_cert_uri_; }
+  const std::string& expectedServerStats() const { return expected_server_stats_; }
+  bool expectSuccess() const { return expect_success_; }
+  Network::Address::IpVersion version() const { return version_; }
+
+protected:
+  TestUtilOptionsBase(const std::string& expected_server_stats, bool expect_success,
+                      Network::Address::IpVersion version)
+      : expected_server_stats_(expected_server_stats), expect_success_(expect_success),
+        version_(version) {}
+
+  void expectedClientCertURI(const std::string& expected_client_cert_uri) {
+    expected_client_cert_uri_ = expected_client_cert_uri;
+  }
+
+private:
+  const std::string expected_server_stats_;
+  const bool expect_success_;
+  const Network::Address::IpVersion version_;
+  std::string expected_client_cert_uri_;
+};
+
+/**
+ * A class to hold the options for testUtil().
+ */
+class TestUtilOptions : public TestUtilOptionsBase {
+public:
+  TestUtilOptions(const std::string& client_ctx_yaml, const std::string& server_ctx_yaml,
+                  const std::string& expected_server_stats, bool expect_success,
+                  Network::Address::IpVersion version)
+      : TestUtilOptionsBase(expected_server_stats, expect_success, version),
+        client_ctx_yaml_(client_ctx_yaml), server_ctx_yaml_(server_ctx_yaml) {}
+
+  const std::string& clientCtxYAML() const { return client_ctx_yaml_; }
+  const std::string& serverCtxYAML() const { return server_ctx_yaml_; }
+
+  TestUtilOptions& expectedClientCertURI(const std::string& expected_client_cert_uri) {
+    TestUtilOptionsBase::expectedClientCertURI(expected_client_cert_uri);
+    return *this;
+  }
+
+  const std::string& expectedClientCertURI() const {
+    return TestUtilOptionsBase::expectedClientCertURI();
```

Comment:
You need to implement this method if it's just the same as in the parent class.

Reply:
@htuch I guess you mean "You need **not** to implement". Currently, it is needed because the setter version of  `expectedClientCertURI()` overrides the version of the parent class, so all the versions of `expectedClientCertURI()` are not inherited and have to be implemented, as far as I understand it.

Anyway, I will rename the setter to be `setExpectedClientCertUri()`, so the getter will be inherited from the parent class.

Reply:
SGTM